### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by icebox827**
- Create the game_logic branch
- Remove mocked logic
- Create game logic

### CODE REVIEW

**Review by LamiaSristy — CHANGES_REQUESTED** [review]
# Status: Change Required. 🔴 

Hello Team @icebox827   and @oxenprogrammer   🙌 

Great work so far and Your game is working smoothly.👏 👏 👏 👏 

However, there are some basic works that remain to get approval and move forward. 🛠️

Please check my points here: 👇

- [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x] The project uses at least two classes (Usually it uses three: Game, Board and Player). The code of the project should be logically divided between classes. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/classes.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))


- [x] If it's an invalid input - You should inform the player about it and asks for their move again

Please make the changes and ask for a re-review.

Feel free to reach me at @lamiasristy 🙋

File: lib/games.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+  def players
+    puts 'enter your name player 1'
+    @player1 = gets.chomp
+    puts "welcome #{@player1}"
+    puts 'enter your name player 2'
+    @player2 = gets.chomp
+    puts "welcome #{@player2}"
+    @player1_token = %w[X O].sample
+    @player2_token = if @player1_token == 'X'
```
Comment (LamiaSristy) [inline lib/games.rb]: No `puts/print` and `gets.chomp` outside `main.rb` file.

**Review by BrianSammit — CHANGES_REQUESTED** [review]
REQUIRED CHANGES :recycle: 
--------------
Hello team!! :wave: excellent job working on the changes.
However, you have to work on some little changes.

------------------
- [x] when I choose a move that it's already taken it works perfectly but it doesn't tell me what's the problem.
----------------------
Kindly fix this and send it for another review :muscle: 
HAPPY CODING!! 

MB5W4D

File: bin/main.rb [hunk 2]
```diff
@@ -1,84 +1,106 @@
...
-        counter += 1
-        board_display(board)
-      end
+      puts 'Move already taken! you loose your turn'
+    end
+    if win?(@board, @player1_token)
+      puts "#{@player1} WINS!!!!"
+      break
+    end
+    unless @board.slice(1, 10).include? ' '
```
Comment (rindrajosia) [inline bin/main.rb]: Move this logic in your lib folder like your `win` method.

File: lib/player.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      true
+    elsif board[1] == player_token and board[5] == player_token and board[9] == player_token
+      true
+    elsif board[3] == player_token and board[5] == player_token and board[7] == player_token
+      true
+    elsif board[1] == player_token and board[4] == player_token and board[7] == player_token
+      true
+    else
+      false
+    end
```
Comment (rindrajosia) [inline lib/player.rb]: Please, find a better way to solve this. Think of DRY code.

**Review by rindrajosia — CHANGES_REQUESTED** [review]
**Status: Changes Required**
Hello team,
Great work on the project so far, but we have to make a few more changes to get this milestone Approved. Let me break down the changes for you.
[REQUIRED]
_GAME BEHAVIOR:_
- [x] If the player chose an occupied position, the game informs the player about it and **asks for their move again.**
- [x] If the player chose an invalid input, the game informs the player about it and **asks for their move again.**

[REQUIRED]
_Implementation details_
- [x] The project uses **at least two classes**. The code of the project should be logically divided between classes. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/classes.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)).
- [x] DRY rule should be respected. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/dry.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)).

Keep up with the efforts towards this project, you are almost there.
Happy coding!!!!
Rindra Josia

**Review by mke2111 — CHANGES_REQUESTED** [review]
# **STATUS: CHANGES REQUIRED**

## Hey Team, I applaud you all for your tremendous efforts towards this milestone. Keep the energy high.

### **REQUIRED CHANGES:**

- [x] remove the .gitkeep file in the lib folder

- [x] In the PlayerMove class, you need to refactor that conditional statement. Keeping your code dry will ensure that you have the best code practices and as such, you won't need to disable Robocop. This will still help in your next projects. 

- You can have the entire winning combinations in an array say 
` [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], .............]` 
- Then search for any winning combination from the user tokens with methods of your choice. (`.any or .all or .include?`)


Make the changes and submit for another review.

## Cheers!!

**Reach me on slack (Roy Mukuye)**

File: lib/move.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+# rubocop:disable Metrics/AbcSize
+
+class PlayerMove
+  def win?(board, player_token)
+    if board[1] == player_token and board[2] == player_token and board[3] == player_token
```
Comment (mke2111) [inline lib/move.rb]: Don't Repeat Yourself.

Comment (oxenprogrammer) [reply lib/move.rb]: refactored as much as possible

**Review by MahmoudBakr23 — APPROVED** [review]
# **Approved ✔✔**

<hr>

**Great job you've done on this project, congratulations! You can now merge and submit.**

<hr>

## **[OPTIONAL]**
The following will complete your project generally and your game specifically as a professional.

- In your main.rb file, please remove the comment in line 43 and the line 55 so that the game stops asking player 2 for his move twice!
- In upcoming projects, try to make your commits more descriptive and also your pull request body.
- In player.rb file the win? method could be refactored and minimized as a previous TSE suggested earlier, try to refactor it so that you follow DRY rules.

<hr>

Good luck in the next projects and happy coding 👨‍💻💻

**Pull request merged**